CASSANDRA-20709: Get SAI MemtableIndex refs before SSTableIndex refs at query time

michaeljmarshall opened the pull request:
[CASSANDRA-20709](https://issues.apache.org/jira/projects/CASSANDRA/issues/CASSANDRA-20709)

SAI predicate search currently has a bug that could result in missing rows due to a concurrent flush during a query. The new test created in this PR shows the point of failure. The problem is that we get the sstable index references before getting the memtable index references. Note that we do it in the correct order in the `ANN OF` query path, but not in the `WHERE` query path.

I propose that we update the QueryView object to hold references to the appropriate memtable indexes.

I removed the problematic search methods from `MemtableIndexManager` because they leave open the possibility for future misuse, in my opinion.

michaeljmarshall commented on src/java/org/apache/cassandra/index/sai/plan/QueryController.java:
```diff
@@ -395,15 +407,15 @@ private KeyRangeIterator getTopKRows(List<PrimaryKey> rawSourceKeys, RowFilter.E
-    private KeyRangeIterator createRowIdIterator(Pair<Expression, Collection<SSTableIndex>> indexExpression)
+    private KeyRangeIterator createRowIdIterator(QueryViewBuilder.QueryExpressionView indexExpression)
     {
-        List<KeyRangeIterator> subIterators = indexExpression.right
+        List<KeyRangeIterator> subIterators = indexExpression.sstableIndexes
                            .stream()
                            .map(index ->
                                 {
                                     try
                                     {
-                                        List<KeyRangeIterator> iterators = index.search(indexExpression.left, mergeRange, queryContext);
+                                        List<KeyRangeIterator> iterators = index.search(indexExpression.expression, mergeRange, queryContext);
                                         // concat the result from multiple segments for the same index
                                         return KeyRangeConcatIterator.builder(iterators.size()).add(iterators).build();
```
Note for reviewers: we could technically remove this `createRowIdIterator` and collapse the logic in the calling method to use a `IndexSearchResultIterator` now. I didn't do that here though because this logic uses `KeyRangeConcatIterator` to more efficiently join sstable index segment search results. Since that is an optimization, I didn't want to update the logic in `IndexSearchResultIterator` on this bug fix. (I'd propose that the real fix is to update `SSTableIndex#search`'s return type from `List<KeyRangeIterator>` to `KeyRangeIterator` and then the implementation can use the `KeyRangeConcatIterator`.)

ekaterinadimitrova2 replied:
How about a comment or TODO, so the next person working in this area is aware of this?

maedhroz commented:
Quick checkstyle run:

```
checkstyle:
    [mkdir] Created dir: /Users/maedhroz/Projects/Apache/asf-cassandra/build/checkstyle
[checkstyle] Running Checkstyle 10.12.1 on 2495 files
[checkstyle] [ERROR] /Users/maedhroz/Projects/Apache/asf-cassandra/src/java/org/apache/cassandra/index/sai/memory/MemtableIndexManager.java:25:8: Unused import - java.util.List. [UnusedImports]
[checkstyle] [ERROR] /Users/maedhroz/Projects/Apache/asf-cassandra/src/java/org/apache/cassandra/index/sai/memory/MemtableIndexManager.java:35:8: Unused import - org.apache.cassandra.db.PartitionPosition. [UnusedImports]
[checkstyle] [ERROR] /Users/maedhroz/Projects/Apache/asf-cassandra/src/java/org/apache/cassandra/index/sai/memory/MemtableIndexManager.java:39:8: Unused import - org.apache.cassandra.dht.AbstractBounds. [UnusedImports]
[checkstyle] [ERROR] /Users/maedhroz/Projects/Apache/asf-cassandra/src/java/org/apache/cassandra/index/sai/memory/MemtableIndexManager.java:40:8: Unused import - org.apache.cassandra.index.sai.QueryContext. [UnusedImports]
[checkstyle] [ERROR] /Users/maedhroz/Projects/Apache/asf-cassandra/src/java/org/apache/cassandra/index/sai/memory/MemtableIndexManager.java:42:8: Unused import - org.apache.cassandra.index.sai.plan.Expression. [UnusedImports]
[checkstyle] [ERROR] /Users/maedhroz/Projects/Apache/asf-cassandra/src/java/org/apache/cassandra/index/sai/memory/MemtableIndexManager.java:43:8: Unused import - org.apache.cassandra.index.sai.iterators.KeyRangeIterator. [UnusedImports]
[checkstyle] [ERROR] /Users/maedhroz/Projects/Apache/asf-cassandra/src/java/org/apache/cassandra/index/sai/memory/MemtableIndexManager.java:44:8: Unused import - org.apache.cassandra.index.sai.iterators.KeyRangeUnionIterator. [UnusedImports]
[checkstyle] [ERROR] /Users/maedhroz/Projects/Apache/asf-cassandra/src/java/org/apache/cassandra/index/sai/memory/MemtableIndexManager.java:45:8: Unused import - org.apache.cassandra.index.sai.utils.PrimaryKey. [UnusedImports]
[checkstyle] [ERROR] /Users/maedhroz/Projects/Apache/asf-cassandra/src/java/org/apache/cassandra/index/sai/plan/QueryController.java:63:8: Unused import - org.apache.cassandra.index.sai.memory.MemtableIndex. [UnusedImports]
[checkstyle] [ERROR] /Users/maedhroz/Projects/Apache/asf-cassandra/src/java/org/apache/cassandra/index/sai/plan/QueryController.java:69:8: Unused import - org.apache.cassandra.utils.Pair. [UnusedImports]
```

maedhroz commented on test/unit/org/apache/cassandra/index/sai/plan/FlushIndexWhileQueryingTest.java:
```diff
@@ -0,0 +1,85 @@
+    {
+        createTable("CREATE TABLE %s (k text PRIMARY KEY, x int)");
+
+        createIndex("CREATE CUSTOM INDEX ON %s(x) USING 'StorageAttachedIndex'");
```
nit: Possible to write this now as `CREATE INDEX ON %s(x) USING 'sai'`

maedhroz commented on test/unit/org/apache/cassandra/index/sai/plan/FlushIndexWhileQueryingTest.java:
```diff
@@ -0,0 +1,85 @@
+        createTable("CREATE TABLE %s (k text PRIMARY KEY, x int)");
+
+        createIndex("CREATE CUSTOM INDEX ON %s(x) USING 'StorageAttachedIndex'");
+        waitForTableIndexesQueryable();
```
nit: This *should* already be embedded in `createIndex()` via `waitForIndexQueryable(qualifiedIndexName.left, qualifiedIndexName.right)` IIRC

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/IndexSearchResultIterator.java:
```diff
@@ -51,22 +53,39 @@ private IndexSearchResultIterator(KeyRangeIterator union, Runnable onClose)
 
         if (includeMemtables)
         {
-            KeyRangeIterator memtableIterator = expression.getIndex().memtableIndexManager().searchMemtableIndexes(queryContext, expression, keyRange);
```
We previously created another union inside `searchMemtableIndexes()`, but I guess this is no longer necessary, since we're just adding to the union here that we create from the `subIterators` anyway. Looks good.

maedhroz commented on src/java/org/apache/cassandra/index/sai/memory/MemtableIndexManager.java:
```diff
@@ -134,47 +127,15 @@ public MemtableIndex getPendingMemtableIndex(LifecycleNewTracker tracker)
     {
-        return liveMemtableIndexMap.values().stream().mapToLong(MemtableIndex::writeCount).sum();
+        // Copy the values. Otherwise, we'll only have a view of the map's values which is subject to change.
+        return new ArrayList<>(liveMemtableIndexMap.values());
```
There's one small optimization we've eliminated here, i.e. when the values collection is empty. Should we preserve that path? Something like...

```
Collection<MemtableIndex> memtableIndexes = liveMemtableIndexMap.values();

if (memtableIndexes.isEmpty())
    return Collections.emptySet();
        
// Copy the values. Otherwise, we'll only have a view of the map's values which is subject to change.
return new ArrayList<>(memtableIndexes);
```
The defensive copy seems critical to the correctness of the fix, and there aren't other callers that want to avoid it. I wonder if `getLiveIndexesSnapshot()` or something similar would help on the caller readability side, but feel free to ignore that (really).

michaeljmarshall replied:
Good suggestions. I will do both.

michaeljmarshall committed:
CASSANDRA-20709: Get SAI MemtableIndex refs before SSTableIndex refs at query time

SAI predicate search currently has a bug that could result in missing
rows due to a concurrent flush during a query. The new test created in
this PR shows the point of failure. The problem is that we get the
sstable index references before getting the memtable index references.
Note that we do it in the correct order in the ANN OF query path, but
not in the WHERE query path.

This commit updates the QueryView object to hold references to the
appropriate memtable indexes.

It also removes the problematic search methods from MemtableIndexManager
to prevent future misuse.

maedhroz commented:
Committed as https://github.com/apache/cassandra/commit/a72dcbc86a6f53eaa45f36514d5838c53be2984f
